File: spoj/TLL237.java
package spoj;

import java.util.*;
import java.io.*;

public class TLL237 {

	public static void main(String[] args) throws IOException {
		Scanner in = new Scanner(System.in);
		
		int N = in.nextInt();
		String seq = in.next();
		
		PriorityQueue<Integer> Q = new PriorityQueue<Integer>();
		for (int i = 0; i < N; ++i)
			Q.offer(0);
		
		for (int i = 0; i < seq.length(); ++i) {
			if (seq.charAt(i) == '1') {
				int less = Q.poll();
				++less;
				Q.offer(less);
			} else {
				int less = Q.poll();
				int less2 = Q.poll();
				++less2;
				Q.offer(less);
				Q.offer(less2);
			}
		}
		
		int less = Q.poll();
		int less2 = Q.poll();
		System.out.println(less + " " + less2);
		
		in.close();
		System.exit(0);
	}
}

File: spoj/TLL237.cpp
#include <iostream>
#include <string>
#include <queue>

using namespace std;

int main() {
    int N;
    cin >> N;
    string seq;
    cin >> seq;
    
    priority_queue<int> Q;
    for (int i = 0; i < N; ++i)
        Q.push(0);
    
    for (int i = 0; i < (int)seq.length(); ++i) {
        if (seq.at(i) == '1') {
            int less = -Q.top();
            Q.pop();
            ++less;
            Q.push(-less);
        } else {
            int less = -Q.top();
            Q.pop();
            int less2 = -Q.top();
            Q.pop();
            ++less2;
            Q.push(-less);
            Q.push(-less2);
        }
    }
    int less = -Q.top();
    Q.pop();
    int less2 = -Q.top();
    
    cout << less << " " << less2 << endl;
    
    return 0;
}
